doc/syntax-philosophy.md

```diff
@@ -0,0 +1,276 @@
+| := +=                 | assignment, accumulation                 | None          |
+| :                     | "has type" with a binder on the left     | Right         |
+| =                     | let binding, field assignment            | Left          |
+| ?                     | separator for labeled rows               | Left          |
+| ,                     | separator for tuples, records, others    | Right         |
+| &                     | separator for record types               | Right         |
+| \|                    | separator for variants and variant types | Left          |
+|-----------------------|------------------------------------------|---------------|
+
+Why this particular precedence table?  The general guiding principle is to
+reduce surprise and grouping parentheses.  Reducing surprise means following
+conventional precedence orders when they exist.  Reducing parentheses means that
+operator A should bind tighter than operator B if it's more typical to pass the
+result of A into B than vice versa (in particular, if A is ill-typed on the
+result of B).  Those desiderata actually constrain the partial order on
+operators quite a lot; the presented table is a consistent total order.
+
+Specifially, we want the following things:
+
+1. Numeric arithmetic has a strong conventional order:
+
+  - Exponentiation is tightest (but Dex has no operator for it)
+
+  - Then unary +, -
+
+  - Then multiplication and division
+
+  - We include .* and *. for scalar-vector multiplication, and make it looser than
+    scalar-scalar multiplication so that 2 * 3 .* [1, 2, 3] needs no parens
+
+  - We also include .** and **. for vector-matrix multiplication, and make it
+    looser yet to continue the pattern of increasing rank (also, 3 .* [1, 2, 3]
+    .** matrix will choose the more efficient order this way, folding the 3 into
+    the vector instead of computing mat-vec and folding the 3 into the result)
+
+  - Finally, we also include ** for matrix-matrix multiplication, and make it
+    looser still, on the same logic
+
+  - Then addition and subtraction
+
+  - The dash-pipe `-|` is a subtraction operator, namely subtracting natural
+    numbers clamping the result to 0
+
+2. Comparison operators are looser than numeric operators because the former can
+   consume the results of the latter, but not vice versa.
+
+  - `~~` is tightest of these because it only makes sense on (structures of)
+    floating-point numbers
+
+  - The standard comparisons `>`, `<`, `>=`, `<=` are given the same precedence
+    by convention
+
+  - The equality operators `==` and `!=` get the next precedence, looser than
+    comparison because `4 < 5 == True` is a reasonable thing to write (if a bit
+    redundant)
+
+3. Boolean arithmetic follows, because it can consume the results of comparisons
+
+  - `&&` is tighter than `||` because the or-of-ands form for Boolean formulas
```

Comment:
Wouldn't it be better to make them have the same precedence and error if not explicitly parenthesized?

Reply:
No?  `2 < x && x < 5 || x == 7` seems like a perfectly sensible expression?

Reply:
I also read `&&` like multiplication and `||` like addition. 🤷

Reply:
I feel like it's better to be explicit in cases like this one, but I guess this precedence is common in other languages, so I won't veto it.